File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.cpp
#include <string>
#include <vector>
#include <queue>
using namespace std;

int arr[101][101];
int visited[101][101][2];
int dx[4] = {0,0,1,-1};
int dy[4] = {1,-1,0,0};

queue<pair<pair<int,int>,int>>q;
int check=0;
int ans = -1;

void bfs(int x,int y, int n, int m){
    q.push({{x,y},0});
    visited[x][y][0]=1;
    while(!q.empty()){
        int kx = q.front().first.first;
        int ky = q.front().first.second;
        int kk = q.front().second;
        q.pop();
        for(int i=0;i<4;i++){
            int nx = kx+dx[i];
            int ny = ky+dy[i];
            int nk = kk;
            if(nx<0 || ny<0|| nx>=n || ny>=m) continue;
            if(arr[nx][ny]==2)  nk =1;
            if(arr[nx][ny]==1||visited[nx][ny][nk]) continue;
            visited[nx][ny][nk]= visited[kx][ky][kk]+1;
            if(arr[nx][ny]==3){
                if(kk==1){
                    ans = visited[kx][ky][kk]+1;
                    return;
                }
            }
            q.push({{nx,ny},nk});
        }
    }
}

int solution(vector<string> maps) {
    int si=0,sj=0;
    for(int i=0;i<maps.size();i++){
        for(int j=0;j<maps[i].size();j++){
            if(maps[i][j]=='S'){
                si = i;
                sj = j;
            }else if(maps[i][j]=='O'){
                arr[i][j]=0;
            }else if(maps[i][j]=='L'){
                arr[i][j]=2;
            }else if(maps[i][j]=='E'){
                arr[i][j]=3;
            }else if(maps[i][j]=='X'){
                arr[i][j]=1;
            }
        }
    }
    
    bfs(si,sj,maps.size(),maps[0].size());
    if(ans==-1){
        return ans;
    }else{
        return ans-1;
    }
}

File: 프로그래머스/2/159993. 미로 탈출/미로 탈출.java
import java.util.*;

class Solution {
    
    private static class Node{
        int x,y;
        int cnt;
        public Node(int x,int y, int cnt){
            this.x=x;
            this.y=y;
            this.cnt=cnt;
        }
    }
    
    private static final int[] dx = {1,-1,0,0};
    private static final int[] dy = {0,0,1,-1};
    //private static boolean[][] visited;
    
    private static int bfs(int x, int y, int n, int m, String[] maps, int check){
        Queue<Node> queue = new ArrayDeque<>();
        queue.add(new Node(x,y,0));
        boolean[][] visited = new boolean[n][m];
        visited[x][y]= true;
        while(!queue.isEmpty()){
            Node curr = queue.poll();
            for(int i=0;i<4;i++){
                int nx = curr.x + dx[i];
                int ny = curr.y + dy[i];
                int nc = curr.cnt + 1;
                if(nx <0 || ny<0|| nx>=n || ny>=m) continue;
                if(visited[nx][ny]) continue;
                char w = maps[nx].charAt(ny);
                if(w=='X') continue;
                visited[nx][ny] =true;
                if((check==1 && w=='L') || (check==2 && w=='E')){
                    //queue.clear();
                    return nc;
                }else{
                    queue.add(new Node(nx,ny,nc));
                }
            }
        }
        return 0;
    }
    
    
    public int solution(String[] maps) {
        int answer = 0;
        
        int si=0, sj=0;
        int li=0, lj=0;
        int n = maps.length;
        int m = maps[0].length();
        
        for(int i=0;i<n;i++){
            String map = maps[i];
            for(int j=0;j<m;j++){
                if(map.charAt(j)=='S'){
                    si=i;
                    sj=j;
                }else if(map.charAt(j)=='L'){
                    li = i;
                    lj = j;
                }
            }
        }
        
        int s = bfs(si,sj,n,m,maps,1);
        int e = bfs(li,lj,n,m,maps,2);
        
        
        return (s==0 || e==0) ? -1 : s+e;
    }
}